AK/Tests/TestIntrusiveList.cpp

```diff
@@ -73,4 +71,78 @@ TEST_CASE(enumeration)
     }
 }
 
+class IntrusiveRefPtrItem : public RefCounted<IntrusiveRefPtrItem> {
+public:
+    IntrusiveRefPtrItem() = default;
+    IntrusiveListNode<IntrusiveRefPtrItem, RefPtr<IntrusiveRefPtrItem>> m_list_node;
+};
+using IntrusiveRefPtrList = IntrusiveList<IntrusiveRefPtrItem, RefPtr<IntrusiveRefPtrItem>, &IntrusiveRefPtrItem::m_list_node>;
+
+TEST_CASE(intrusive_ref_ptr_no_ref_leaks)
+{
+    auto item = adopt(*new IntrusiveRefPtrItem());
+    EXPECT_EQ(1u, item->ref_count());
+    IntrusiveRefPtrList ref_list;
+
+    ref_list.append(*item);
+    EXPECT_EQ(2u, item->ref_count());
+
+    ref_list.remove(*item);
+    EXPECT_EQ(1u, item->ref_count());
+}
+
+TEST_CASE(intrusive_ref_ptr_clear)
+{
+    auto item = adopt(*new IntrusiveRefPtrItem());
+    EXPECT_EQ(1u, item->ref_count());
+    IntrusiveRefPtrList ref_list;
+
+    ref_list.append(*item);
+    EXPECT_EQ(2u, item->ref_count());
+
+    ref_list.clear();
+    EXPECT_EQ(1u, item->ref_count());
+}
+
+TEST_CASE(intrusive_ref_ptr_destructor)
+{
+    auto item = adopt(*new IntrusiveRefPtrItem());
+    EXPECT_EQ(1u, item->ref_count());
+
+    {
+        IntrusiveRefPtrList ref_list;
+        ref_list.append(*item);
+        EXPECT_EQ(2u, item->ref_count());
+    }
+
+    EXPECT_EQ(1u, item->ref_count());
+}
+
+// There is currently a bug in the NonnullRefPtr specialization of IntrusiveList,
+// so this test fails to compile at the moment. Including it for future use.
```

Comment:
:thinking:

How does this fail?
I do remember checking that it compiles, though not with a lot of method calls...

Reply:
Maybe I'm holding it wrong? 

```
inja: Entering directory `/home/bgianf/src/serenity/Build/i686'
[0/2] Re-checking globbed directories...
[2/3] Building CXX object AK/Tests/CMakeFiles/TestIntrusiveList.dir/TestIntrusiveList.cpp.o
FAILED: AK/Tests/CMakeFiles/TestIntrusiveList.dir/TestIntrusiveList.cpp.o
/usr/bin/ccache /home/bgianf/src/serenity/Toolchain/Local/i686/bin/i686-pc-serenity-g++ -DDEBUG -DENABLE_COMPILETIME_FORMAT_CHECK -DSANITIZE_PTRS -I../../Userland/Libraries -I../../. -I. -I../../Userland/Libraries/LibC -I../../Userland/Libraries/LibM -I../../Userland/Libraries/LibSystem -I../../Userland/Services -I../../Userland -IUserland/Services -IUserland/Libraries -IUserp
In file included from ../../AK/Tests/TestIntrusiveList.cpp:27:
../.././AK/IntrusiveList.h: In instantiation of 'class AK::IntrusiveList<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem>, &IntrusiveNonnullRefPtrItem::m_list_node>':
../../AK/Tests/TestIntrusiveList.cpp:135:32:   required from here
../.././AK/IntrusiveList.h:352:7: error: could not convert template argument '&IntrusiveNonnullRefPtrItem::m_list_node' from 'AK::IntrusiveListNode<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem> > IntrusiveNonnullRefPtrItem::*' to 'AK::IntrusiveListNode<IntrusiveNonnullRefPtrItem, AK::RefPtr<IntrusiveNonnullRefPtrItem, AK::RefPtrTraits<IntrusiveNonn'
  352 | class IntrusiveList<T, NonnullRefPtr<T>, member> : public IntrusiveList<T, RefPtr<T>, member> {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../AK/Tests/TestIntrusiveList.cpp: In function 'void __test_intrusive_nonnull_ref_ptr_intrusive()':
../../AK/Tests/TestIntrusiveList.cpp:137:22: error: 'using IntrusiveNonnullRefPtrList = class AK::IntrusiveList<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem>, &IntrusiveNonnullRefPtrItem::m_list_node>' {aka 'class AK::IntrusiveList<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem>, &IntrusiveNonnullRefPtrItem::m_list_node>'}'
  137 |     nonnull_ref_list.append(*item);
```

Reply:
Ohhh it needs to entirely rewrite `IntrusiveListNode<T, NonnullRefPtr<T>>` to `IntrusiveListNode<T, RefPtr<T>>`, not just subclass it
I guess I made the node hold a `RefPtr<T>` when I tested it...

It'll get a little ugly, but should be fixable...

Reply:
I made a PR on your repo to this branch with a fix :^)

Reply:
...which apparently doesn't fix anything...

Reply:
:) Did you try un-ifdefing that test?

Reply:
hmm...weird, it works for me, I wonder if it's a clang thing?

Reply:
I `#ifdef`'d around clang, hopefully it builds and you can un-if0 your other test.

Reply:
Do you  just want to send this as a separate PR?

I don't need NonnullRefPtr here for any reason, just wanted to add a test to cover it.
This shouldn't be gated on that working?

Reply:
Either way is fine by me, I can add this fix and un-if0 the test (or just add it) in a new PR.

Reply:
I would prefer that, thanks!